File: android/network/src/main/kotlin/com/ridelink/network/security/TlsControlChannel.kt
package com.ridelink.network.security

import com.ridelink.core.model.SpkiHash
import com.ridelink.core.protocol.Sas
import com.ridelink.core.security.IdentityCertificate
import com.ridelink.network.control.ChannelSecurity
import com.ridelink.network.control.ControlChannel
import com.ridelink.network.control.ControlListener
import com.ridelink.network.control.ControlSocket
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.KeyManager
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLEngine
import javax.net.ssl.SSLServerSocket
import javax.net.ssl.SSLSocket
import javax.net.ssl.TrustManager
import javax.net.ssl.X509ExtendedTrustManager

/** PROTOCOL §1 / ADR-007: TLS 1.3 and nothing else. Not a preference — the only enabled protocol. */
private const val TLS_1_3 = "TLSv1.3"

/**
 * The three TLS 1.3 cipher suites. Used to assert the negotiated connection really is TLS 1.3.
 *
 * That check is deliberately not `SSLSession.getProtocol() == "TLSv1.3"`: the Phase 1b spike
 * measured Conscrypt's **server-side** session reporting `TLSv1.2` on a connection where only
 * TLS 1.3 was ever enabled and a TLS-1.3-only suite was negotiated. Asserting on the protocol
 * string would reject good sessions. The suite name cannot lie in the same way, because these
 * three exist only in TLS 1.3 (ADR-018 §3, `docs/test-results/phase1b-security-spike-20260827.md`).
 */
private val TLS_1_3_CIPHER_SUITES =
    setOf("TLS_AES_128_GCM_SHA256", "TLS_AES_256_GCM_SHA384", "TLS_CHACHA20_POLY1305_SHA256")

/** How long a TLS handshake may take before the socket is abandoned. */
private const val HANDSHAKE_TIMEOUT_MS = 5_000

/** Matches the Phase 1a `ControlSocket.connect` timeout, and iOS's `connectTimeoutMs`. */
private const val CONNECT_TIMEOUT_MS = 5_000

/**
 * Which TLS 1.3 implementation provides the sockets, and how its keying-material exporter
 * (RFC 8446 §7.5) is reached.
 *
 * **One interface, one production implementation, and a specific reason for the seam.** Android's
 * default JSSE provider *is* Conscrypt-over-BoringSSL and reaches the exporter through
 * `android.net.ssl.SSLSockets` — a class that does not exist on a plain JVM. A desktop JVM's
 * default provider (SunJSSE) speaks TLS 1.3 but exposes **no exporter at all**, which is a gap in
 * the JDK, not in this design. Without this seam the entire secure control plane would be testable
 * only on a phone, and the SAS — the one thing a user is asked to check with their own eyes —
 * would have no laptop test.
 *
 * The test implementation supplies Conscrypt directly, so what the JVM suite exercises is the same
 * TLS stack Android ships and the same `exportKeyingMaterial` the platform class delegates to, one
 * call frame lower. That substitution is stated plainly in
 * `docs/test-results/phase1b-security-spike-20260827.md` and is closed by the real-device gate,
 * not by this comment.
 */
interface TlsProvider {
    fun sslContext(
        keyManagers: Array<KeyManager>,
        trustManagers: Array<TrustManager>,
        random: SecureRandom,
    ): SSLContext

    /** Returns null when the exporter is unavailable; callers must treat that as a hard failure. */
    fun exportKeyingMaterial(
        socket: SSLSocket,
        label: String,
        context: ByteArray,
        length: Int,
    ): ByteArray?
}

/**
 * The production provider: the platform's own TLS 1.3 (Conscrypt on Android) plus
 * `android.net.ssl.SSLSockets.exportKeyingMaterial`, public SDK since **API 31** — exactly the
 * ADR-011 `minSdk`, verified against `api-versions.xml` rather than assumed (ADR-018 §1).
 */
object AndroidTlsProvider : TlsProvider {
    override fun sslContext(
        keyManagers: Array<KeyManager>,
        trustManagers: Array<TrustManager>,
        random: SecureRandom,
    ): SSLContext = SSLContext.getInstance(TLS_1_3).apply { init(keyManagers, trustManagers, random) }

    override fun exportKeyingMaterial(
        socket: SSLSocket,
        label: String,
        context: ByteArray,
        length: Int,
    ): ByteArray? =
        runCatching { android.net.ssl.SSLSockets.exportKeyingMaterial(socket, label, context, length) }
            .getOrNull()
}

/**
 * The channel-security facts for one completed TLS 1.3 handshake.
 *
 * Note what is *not* here: no method returns the peer's certificate, its subject, or anything else
 * a caller could accidentally start trusting. Trust is [peerIdentitySpkiSha256] and nothing else
 * (ADR-012).
 */
internal class TlsChannelSecurity(
    private val socket: SSLSocket,
    peerCertificate: X509Certificate,
    override val peerCertificateStructurallyValid: Boolean,
    private val provider: TlsProvider,
) : ChannelSecurity {
    override val peerIdentitySpkiSha256: SpkiHash =
        IdentityCertificate.spkiHashOfEncoded(peerCertificate.publicKey.encoded)

    override val negotiatedCipherSuite: String = socket.session.cipherSuite

    /**
     * PROTOCOL §4.5.1 / ADR-018. The exporter output never leaves this method except as six
     * decimal digits, and neither the digits nor the 32 bytes have any log path
     * (ARCHITECTURE §11).
     */
    override fun deriveSas6(): String? {
        val exported = provider.exportKeyingMaterial(socket, Sas.EXPORTER_LABEL, EMPTY_CONTEXT, Sas.EXPORTER_LENGTH_BYTES)
        return exported?.let(Sas::deriveSas6)
    }

    private companion object {
        /**
         * ADR-018 §2: TLS 1.3 always hashes a context value, so an absent context and an empty one
         * are the same input — measured equal on Conscrypt, and the only form Apple's public API
         * can express. Written as an explicit empty array rather than `null` so the choice reads
         * as deliberate.
         */
        val EMPTY_CONTEXT = ByteArray(0)
    }
}

/**
 * **The production control channel.** TCP + TLS 1.3, mutually authenticated with the device's own
 * self-signed identity certificate (ADR-007, ADR-017).
 *
 * Both ends require a peer certificate. A one-sided handshake would leave the connecting side's
 * identity unbound, which would make the SPKI pin on that side meaningless — the peer would be
 * pinning a key nobody proved possession of.
 *
 * PKI validation is deliberately absent: there is no CA, no hostname to check, and
 * [PinningTrustManager] therefore accepts every certificate at the *transport* layer and records
 * it. Trust is applied one layer up, by [com.ridelink.core.security.PeerTrust], against the
 * pinned `identity_spki_sha256`. That split is what ARCHITECTURE §4.3 describes, and it is why
 * accepting here is not a hole: nothing is trusted until the pin says so, and a `TrustManager`
 * cannot express "ask the user" anyway.
 */
class TlsControlChannel(
    private val identity: DeviceIdentity,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
    private val provider: TlsProvider = AndroidTlsProvider,
    private val secureRandom: SecureRandom = SecureRandom(),
) : ControlChannel {
    override val transportLabel: String = "TLS 1.3 / MUTUAL / SPKI-PINNED"
    override val isSecure: Boolean = true

    private val sslContext: SSLContext by lazy {
        provider.sslContext(identity.keyManagers(), arrayOf(PinningTrustManager()), secureRandom)
    }

    override suspend fun bind(): ControlListener =
        withContext(ioDispatcher) {
            val server = sslContext.serverSocketFactory.createServerSocket() as SSLServerSocket
            server.reuseAddress = true
            server.enabledProtocols = arrayOf(TLS_1_3)
            server.needClientAuth = true // both peers prove key possession, or neither pin means anything
            server.bind(InetSocketAddress(0)) // dynamic port; advertised via mDNS by the caller
            ControlListener(server) { socket -> accept(socket) }
        }

    private suspend fun accept(server: ServerSocket): ControlSocket =
        withContext(ioDispatcher) {
            val socket = server.accept() as SSLSocket
            finishHandshake(socket, isInitiator = false)
        }

    override suspend fun connect(
        host: String,
        port: Int,
    ): ControlSocket =
        withContext(ioDispatcher) {
            // Connect the TCP layer with an explicit timeout first, then start TLS on top: an
            // SSLSocketFactory.createSocket(host, port) would connect with no timeout at all.
            val plain = Socket()
            plain.connect(InetSocketAddress(host, port), CONNECT_TIMEOUT_MS)
            val socket =
                runCatching {
                    sslContext.socketFactory.createSocket(plain, host, port, true) as SSLSocket
                }.getOrElse { failure ->
                    runCatching { plain.close() }
                    throw failure
                }
            socket.useClientMode = true
            finishHandshake(socket, isInitiator = true)
        }

    private fun finishHandshake(
        socket: SSLSocket,
        isInitiator: Boolean,
    ): ControlSocket {
        ControlSocket.configureTcp(socket)
        socket.enabledProtocols = arrayOf(TLS_1_3)
        try {
            // soTimeout bounds the handshake so an unresponsive peer cannot pin this coroutine
            // forever; it is cleared afterwards because the read loop must block indefinitely and
            // let the application PING/PONG (PROTOCOL §1) decide liveness instead.
            socket.soTimeout = HANDSHAKE_TIMEOUT_MS
            socket.startHandshake()
            socket.soTimeout = 0

            check(socket.session.cipherSuite in TLS_1_3_CIPHER_SUITES) {
                "negotiated ${socket.session.cipherSuite}, which is not a TLS 1.3 cipher suite"
            }
            val peerCertificate =
                socket.session.peerCertificates.firstOrNull() as? X509Certificate
                    ?: error("peer presented no X.509 certificate")

            return ControlSocket(
                socket = socket,
                isInitiator = isInitiator,
                ioDispatcher = ioDispatcher,
                security =
                    TlsChannelSecurity(
                        socket = socket,
                        peerCertificate = peerCertificate,
                        peerCertificateStructurallyValid = isStructurallyValid(peerCertificate),
                        provider = provider,
                    ),
            )
        } catch (failure: Throwable) {
            runCatching { socket.close() }
            throw failure
        }
    }

    /**
     * The structural half of PROTOCOL §4.1's certificate check: well-formed, self-signature
     * verifies, inside its validity window. Explicitly **not** a chain or hostname check — there
     * is no CA and no name to check.
     *
     * A false here becomes `ERROR/certificate_invalid`, which is a distinct code from
     * `pin_mismatch` precisely so a phone with a wrong clock is not reported to its owner as an
     * attack (ADR-012).
     */
    private fun isStructurallyValid(certificate: X509Certificate): Boolean =
        runCatching {
            certificate.checkValidity()
            certificate.verify(certificate.publicKey) // self-signed: it is its own issuer
            true
        }.getOrDefault(false)

    /**
     * Accepts every certificate at the transport layer and lets the SPKI pin decide, one layer up.
     *
     * `X509ExtendedTrustManager` rather than the plain interface: JSSE calls the socket/engine
     * overloads when they exist, and a subclass that only overrides the two-argument forms would
     * silently not be consulted.
     */
    private class PinningTrustManager : X509ExtendedTrustManager() {
        override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) = Unit

        override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) = Unit

        override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?, socket: Socket?) = Unit

        override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?, socket: Socket?) = Unit

        override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?, engine: SSLEngine?) = Unit

        override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?, engine: SSLEngine?) = Unit

        override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
    }
}

File: ios/Packages/RideLinkPlatform/Sources/RideLinkPlatform/Security/TlsControlChannel.swift
import Foundation
import Network
import RideLinkCore
import Security

/// The channel-security facts for one completed TLS 1.3 handshake.
///
/// Note what is *not* here: nothing returns the peer's certificate, its subject, or anything else
/// a caller could accidentally start trusting. Trust is `peerIdentitySpkiSha256` and nothing else
/// (ADR-012).
struct TlsChannelSecurity: ChannelSecurity, @unchecked Sendable {
    let peerIdentitySpkiSha256: SpkiHash
    let peerCertificateStructurallyValid: Bool
    let negotiatedProtocolDescription: String
    private let metadata: sec_protocol_metadata_t

    init(
        peerIdentitySpkiSha256: SpkiHash,
        peerCertificateStructurallyValid: Bool,
        negotiatedProtocolDescription: String,
        metadata: sec_protocol_metadata_t
    ) {
        self.peerIdentitySpkiSha256 = peerIdentitySpkiSha256
        self.peerCertificateStructurallyValid = peerCertificateStructurallyValid
        self.negotiatedProtocolDescription = negotiatedProtocolDescription
        self.metadata = metadata
    }

    /// PROTOCOL §4.5.1 / ADR-018.
    ///
    /// Uses the **context-less** exporter call. That is not a shortcut: Apple's public API has no
    /// way to pass a present-but-zero-length context — `sec_protocol_metadata_create_secret_with_context`
    /// returns nil for `context_len: 0` — and under TLS 1.3 the two are the same input anyway,
    /// because RFC 8446 §7.5 always hashes a context value. Measured equal against Conscrypt and
    /// OpenSSL in `docs/test-results/phase1b-security-spike-20260827.md`.
    ///
    /// The exporter output never leaves this method except as six decimal digits, and neither the
    /// digits nor the 32 bytes have any log path (ARCHITECTURE §11).
    func deriveSas6() -> String? {
        let label = Array(Sas.exporterLabel.utf8)
        let exported: dispatch_data_t? = label.withUnsafeBufferPointer { buffer in
            guard let base = buffer.baseAddress else { return nil }
            return sec_protocol_metadata_create_secret(
                metadata,
                label.count,
                UnsafeRawPointer(base).assumingMemoryBound(to: CChar.self),
                Sas.exporterLengthBytes
            )
        }
        guard let exported else { return nil }
        var bytes = [UInt8]()
        (exported as DispatchData).enumerateBytes { buffer, _, _ in bytes.append(contentsOf: buffer) }
        guard bytes.count >= 4 else { return nil }
        return Sas.deriveSas6(bytes)
    }
}

/// **The production control channel.** TCP + TLS 1.3, mutually authenticated with the device's own
/// self-signed identity certificate (ADR-007, ADR-017).
///
/// Both ends require a peer certificate
/// (`sec_protocol_options_set_peer_authentication_required`). A one-sided handshake would leave
/// the connecting side's identity unbound, which would make the SPKI pin on that side meaningless
/// — the peer would be pinning a key nobody proved possession of.
///
/// PKI validation is deliberately absent: there is no CA and no hostname to check, so the verify
/// block accepts the transport and trust is applied one layer up by `RideLinkCore.PeerTrust`
/// against the pinned `identity_spki_sha256`. That split is what ARCHITECTURE §4.3 describes, and
/// it is why accepting here is not a hole: nothing is trusted until the pin says so, and a verify
/// block cannot express "ask the user" anyway.
public struct TlsControlChannel: ControlChannel, @unchecked Sendable {
    public let transportLabel = "TLS 1.3 / MUTUAL / SPKI-PINNED"
    public let isSecure = true

    private let identity: DeviceIdentity
    private let connectTimeoutMs: Int64
    private let handshakeTimeoutMs: Int64 = 5000
    private let verifyQueue = DispatchQueue(label: "com.ridelink.platform.tls.verify")

    public init(identity: DeviceIdentity, connectTimeoutMs: Int64 = 5000) {
        self.identity = identity
        self.connectTimeoutMs = connectTimeoutMs
    }

    public func bind() async throws -> ControlListener {
        try await ControlListener.bind(parameters: makeParameters()) { connection in
            // An accepted connection has not finished its TLS handshake yet, and the peer
            // certificate and exporter do not exist until it has.
            try await connection.awaitReady(timeoutMs: handshakeTimeoutMs)
            try attachSecurity(to: connection)
        }
    }

    public func connect(host: String, port: UInt16) async throws -> ControlConnection {
        let connection = try await ControlConnection.connect(
            host: host,
            port: port,
            parameters: makeParameters(),
            connectTimeoutMs: connectTimeoutMs
        )
        do {
            try attachSecurity(to: connection)
        } catch {
            connection.close()
            throw error
        }
        return connection
    }

    /// Reads the peer certificate and the negotiated parameters off a connection that has already
    /// reached `.ready`, and hands the result to the framing layer as an opaque `ChannelSecurity`.
    ///
    /// Done here rather than in the verify block because a listener's `NWParameters` — and
    /// therefore its verify block — are shared across every inbound connection, while these facts
    /// are per-connection. `sec_protocol_metadata` from the connection itself is the per-connection
    /// view.
    public func attachSecurity(to connection: ControlConnection) throws {
        guard let metadata = connection.securityProtocolMetadata else {
            throw ControlTransportError.connectFailed("no TLS metadata on a connection that should be TLS")
        }
        let version = sec_protocol_metadata_get_negotiated_tls_protocol_version(metadata)
        guard version == .TLSv13 else {
            throw ControlTransportError.connectFailed(
                "negotiated \(Self.describe(version)), but PROTOCOL §1 requires TLS 1.3")
        }
        guard let certificate = Self.peerLeafCertificate(from: metadata) else {
            throw ControlTransportError.connectFailed("peer presented no certificate")
        }
        connection.attachSecurity(
            TlsChannelSecurity(
                peerIdentitySpkiSha256: try PeerCertificateInspector.identitySpkiSha256(of: certificate),
                peerCertificateStructurallyValid: PeerCertificateInspector.isStructurallyValid(certificate),
                negotiatedProtocolDescription: Self.describe(version),
                metadata: metadata
            )
        )
    }

    private func makeParameters() -> NWParameters {
        let tlsOptions = NWProtocolTLS.Options()
        let sec = tlsOptions.securityProtocolOptions
        let localIdentity = sec_identity_create(identity.secIdentity)!

        sec_protocol_options_set_local_identity(sec, localIdentity)
        // Not a floor and a ceiling for negotiation room — this pins TLS 1.3 exactly. NFR-06 and
        // PROTOCOL §1 do not admit a 1.2 fallback, and making it unnegotiable is stronger than
        // detecting it afterwards.
        sec_protocol_options_set_min_tls_protocol_version(sec, .TLSv13)
        sec_protocol_options_set_max_tls_protocol_version(sec, .TLSv13)
        sec_protocol_options_set_peer_authentication_required(sec, true)
        sec_protocol_options_set_challenge_block(sec, { _, complete in
            complete(localIdentity) // the client certificate half of mutual authentication
        }, verifyQueue)
        sec_protocol_options_set_verify_block(sec, { _, _, complete in
            complete(true) // see the type-level comment: trust is the SPKI pin, one layer up
        }, verifyQueue)

        let parameters = NWParameters(tls: tlsOptions)
        if let tcpOptions = parameters.defaultProtocolStack.transportProtocol as? NWProtocolTCP.Options {
            tcpOptions.noDelay = true
        }
        parameters.allowLocalEndpointReuse = true
        return parameters
    }

    private static func peerLeafCertificate(from metadata: sec_protocol_metadata_t) -> SecCertificate? {
        var leaf: SecCertificate?
        sec_protocol_metadata_access_peer_certificate_chain(metadata) { secCertificate in
            guard leaf == nil else { return } // the leaf is the peer's own certificate
            leaf = sec_certificate_copy_ref(secCertificate).takeRetainedValue()
        }
        return leaf
    }

    private static func describe(_ version: tls_protocol_version_t) -> String {
        switch version {
        case .TLSv13: return "TLSv1.3"
        case .TLSv12: return "TLSv1.2"
        case .TLSv11: return "TLSv1.1"
        case .TLSv10: return "TLSv1.0"
        default: return "unknown(\(version.rawValue))"
        }
    }
}
